### PR TITLE
Remove extra attribute in OpMaker

### PR DESCRIPTION
**Pull request by zyfncg**
### PR types

Breaking changes

### PR changes

Others

### 一、背景

`OpMaker`是飞桨框架用于定义一个算子的输入(Input)输出(Output)以及属性(Attribute)信息的基础数据结构，原则上讲一个算子`OpMaker`中定义的参数应该与Python API中的参数保持一致。但由于之前对`OpMaker`没有较强的规范约束，在各种因素的作用下，导致部分算子的`OpMaker`定义中出现了一些不在Python API范围内的参数。以`conv2d`为例，Paddle提供的Python API输入参数共 8 个，而`OpMaker`中定义的输入参数项（Input和Attribute）共 29 个，多了 21 个非用户传入的参数项，多出来的这些参数项大多用于传递特定场景下特殊Kernel的额外配置信息，而非描述卷积算子数学逻辑运算上输入输出参数，这与`OpMaker`设计的职责不符，严格来说属于不规范的使用。

`OpMaker`参数的不规范也会向外进行传递：基于`OpMaker`生成的`OpDesc`是Program（IR表示）的基本单元，不规整的`OpDesc`会导致Program（IR）中出现许多无用的冗余信息，进而引起训练和下游推理等业务的不稳定。

本次`OpMaker`规范化清理升级的意义，是为了让飞桨的Op定义更加简单清晰，便于开发者理解和降低硬件接入的成本，也为后续通过算子Yaml配置直接生成OpMaker铺平了道路；同时也会让模型保存的Program信息更加规范合理，更好的服务于模型的推理和部署。

### 二、目标

1.  清理`OpMaker`中不规范的参数（目前标记为`AsExtra`的Attribute），将其放入到更合理的位置进行管理，核心目标是：OpMaker的定义输入、输出仅保存最基础的信息（理想情况是与 Python API 传入的参数能够保持一致），其余信息放到别的地方进行管理，使`OpMaker`保持规范简洁。
2. 在目标1的基础上，模型保存的Program中只包含`OpMaker`清理后所具有的Attribute（目前标记为Extra的属性会去掉），使保存的Program规范整洁，并且保存的模型能够正常的加载以及执行训练和推理。

### 三、设计

image
出于兼容性考虑`OpMaker`的清理并不是将`OpMaker`中的`Extra`参数直接删除，而是从`OpMaker`中去掉后统一在Yaml配置文件中（paddle/phi/api/yaml/api_compat.yaml）进行管理，并将这类参数作为`OpDesc`和`OperatorBase`的`runtime_attrs`成员变量，在`OpDesc`和`OperatorBase`创建时会将这些在Yaml中配置的参数信息（考虑到性能问题，这里在框架编译时会将Yaml配置信息解析并生成为一个全局静态map存储数据）恢复到这两个类的`runtime_attrs`中，在执行器运行对应Op时，对于Attribute的获取会从`attrs`和`runtime_attrs`中进行查找，这样处理的行为理论上与`OpMaker`清理前是完全能够对齐的，所以`OpMaker`清理在设计上可以保证兼容性。
移入Yaml配置文件中的`Extra`参数后续将随着机制的完善逐步删除，从而最终做到彻底移除。

本PR中以cond2d为例，将`OpMaker`中标记为Extra的Attribute移除后放置在`paddle/phi/api/yaml/api_compat.yaml`中进行管理。后续待机制进一步完善和验证后，将会对更多的算子`OpMaker`进行调整。

English
--
### Ⅰ. Background

`Opmaker` is the basic structure in PaddlePaddle to define the input, output and attribute information of Operator. The parameters defined in `OpMaker` of Operator should be consistent with parameters in its Python API. However, because the lack of strong constraints on `OpMaker`, some parameters in `OpMaker` that are not within the Python API, this is caused by various reasons. 
For example, `conv2d` has 8 parameters in Python API, but the number of input parameters in `cond2d OpMaker` is 29. There are 21 parameters passed by framewrok, and most of these parameters are used to pass additional configuration for special kernels, such as mkldnn kernel, This is inconsistent with the design philosophy of `OpMaker`.

The nonstandard parameters of `OpMaker` will also influence the outside: the `OpDesc` generated by `OpMaker` is the basic unit of program (IR representation), and there will be some useless redundant information in the program (IR) caused by messy `OpMaker`, which will lead to instability of training and inference.

The significance of cleaning and upgrading `OpMaker` is to make the definition of Operato simpler and cleaner. At the same time, it will make the program saved by the model more standardized, and it is also conducive to model inference and deployment.

### Ⅱ. Target
1. Clean up the extra parameters in `OpMaker` (the attribute marked as `AsExtra`) and put them together for management. The final goal is: the Inputs, Outputs  and Attribute of `OpMaker` only include basic math information (the parameters in `OpMaker` should be consistent with Python API). 

2. The program saved by the model only contains the attributes that `OpMaker` has after cleaning (the attributes marked as extra will be removed at present), so that the saved program is standardized and tidy, and then the saved model can be loaded for training and inference.

### Ⅲ. Design
image
The cleanup of `OpMaker` does not delete the `extra` parameter in `OpMaker` directly, instead, we put them in the yaml file (paddle/phi/api/yaml/api_compat.yaml) after removing them from `OpMaker`, and takes these parameters as the `runtime_attrs` of `OpDesc` and `OperatorBase`.
When `OpDesc` and `OperatorBase` are created, the parameter configured in yaml will be restored to the `runtime_attrs`，the parameter in `OpMaker` is still in `attrs`. (In order to better performance, yaml configuration will be parsed and generated into a global static map during code compilation). 
When the executor runs the Operator, the attribute will be searched from `attrs` and `runtime_ attrs`, it's result is same with before, so our design can ensure compatibility.

The extra parameter moved into the yaml file will be deleted gradually with the improvement of the design. Finally they will be deleted completely.

In this PR, we choose `cond2d` as an example, remove the attributes marked as extra in `OpMaker` and put them in `paddle/phi/api/yaml/api_compat.yaml`. After this design is verified, we will clean up more `OpMaker` of Operators in the future.

### CODE REVIEW

File: paddle/fluid/framework/new_executor/interpretercore.cc [hunk 1]
```diff
@@ -998,6 +998,7 @@ void InterpreterCore::Prepare(
     const std::vector<std::string>& feed_names,
     const std::vector<framework::LoDTensor>& feed_tensors,
     bool prepare_feed) {
+  VLOG(1) << "################# InterpreterCore::Prepare";
```
Comment (chenwhql) [inline paddle/fluid/framework/new_executor/interpretercore.cc]: if needless, recommend to remove debug info, or use higher log level

File: paddle/fluid/framework/new_executor/interpretercore.cc [hunk 2]
```diff
@@ -1043,6 +1044,7 @@ void InterpreterCore::Prepare(
   if (prepare_feed) {
     FeedInput();
   }
+  VLOG(1) << "################# InterpreterCore::Prepare Finish";
```
Comment (chenwhql) [inline paddle/fluid/framework/new_executor/interpretercore.cc]: same as above

**Review by chenwhql — DISMISSED** [review]
LGTM

File: paddle/fluid/framework/new_executor/interpretercore.cc (hunk 1 above)
Comment (zyfncg) [reply paddle/fluid/framework/new_executor/interpretercore.cc]: done

File: paddle/fluid/framework/new_executor/interpretercore.cc (hunk 2 above)
Comment (zyfncg) [reply paddle/fluid/framework/new_executor/interpretercore.cc]: done

File: paddle/fluid/framework/op_desc.h [hunk 3]
```diff
@@ -221,6 +225,7 @@ class OpDesc {
   VariableNameMap outputs_;
   // attribute name => all original attrs
   AttributeMap attrs_;
+  AttributeMap runtime_attrs_;
```
Comment (qingqing01) [inline paddle/fluid/framework/op_desc.h:233]: add comments for "runtime_attrs_"

File: paddle/fluid/operators/conv_transpose_op.cc [hunk 4]
```diff
@@ -302,14 +293,6 @@ void Conv3DTransposeOpMaker::Make() {
       "\"SAME\",\"VALID\". Set to \"EXPLICIT\" for explicit padding. "
       "Set to \"SAME\" or \"VALID\" for algorithm of padding. ")
       .SetDefault("EXPLICIT");
-  AddAttr<int>("workspace_size_MB",
```
Comment (qingqing01) [inline paddle/fluid/operators/conv_transpose_op.cc]: Conv2D没删除workspace_size_MB，这里却删除了，保持一致比较好些吧~

File: paddle/fluid/operators/conv_op.cc [hunk 5]
```diff
@@ -348,82 +348,6 @@ void Conv2DOpMaker::Make() {
...
-  AddAttr<std::vector<float>>("Scale_weights",
-                              "Scale_weights to be used for int8 weights data."
-                              "Only used with MKL-DNN INT8.")
-      .SetDefault({1.0f})
-      .AsExtra();
-  AddAttr<bool>("force_fp32_output",
-                "(bool, default false) Force INT8 kernel output FP32, only "
-                "used in MKL-DNN INT8")
-      .SetDefault(false)
-      .AsExtra();
```
Comment (qingqing01) [inline paddle/fluid/operators/conv_op.cc:426]: 没有看到恢复这些属性的单测代码

File: paddle/fluid/framework/ir/mkldnn/conv_bias_mkldnn_fuse_pass.cc [hunk 6]
```diff
@@ -298,6 +298,9 @@ void ConvBiasFusePass::ApplyImpl(ir::Graph* graph) const {
       for (auto& attr : conv->Op()->GetAttrMap()) {
         desc.SetAttr(attr.first, attr.second);
       }
+      for (auto& attr : conv->Op()->GetRuntimeAttrMap()) {
+        desc.SetAttr(attr.first, attr.second);
+      }
```
Comment (Silv3S) [inline paddle/fluid/framework/ir/mkldnn/conv_bias_mkldnn_fuse_pass.cc:303]: Why is it needed only in `conv_bias_mkldnn_fuse_pass` and not in other conv fusions?

**Review by Silv3S — APPROVED** [review]
LGTM

File: paddle/fluid/framework/ir/mkldnn/conv_bias_mkldnn_fuse_pass.cc (hunk 6 above)
Comment (zyfncg) [reply paddle/fluid/framework/ir/mkldnn/conv_bias_mkldnn_fuse_pass.cc:303]: The attribute setting in `conv_bias_mkldnn_fuse_pass` is special.
In other conv fusions, the extra attributes are set by attr name directly, for example:

image

image

File: paddle/fluid/framework/op_desc.h (hunk 3 above)
Comment (zyfncg) [reply paddle/fluid/framework/op_desc.h:233]: done, thx

File: paddle/fluid/operators/conv_transpose_op.cc (hunk 4 above)
Comment (zyfncg) [reply paddle/fluid/operators/conv_transpose_op.cc]: done, thx~

File: paddle/fluid/operators/conv_op.cc (hunk 5 above)
Comment (zyfncg) [reply paddle/fluid/operators/conv_op.cc:426]: 这些属性信息通过Yaml配置恢复后再使用的情况在目前的单测中基本都会覆盖到，如果要新增单测来测试这里参数恢复的情况，成本会比较高：一方面OpMaker中参数删除后，很难通过简单的测试来验证行为上和删除前是否有发生变化，另一方面Op对象的创建有多个接口，在OpTest中要覆盖这些场景的开发成本也比较高。所以综合考虑这里暂时没有新增单测来做专门的测试。
对于模型加载Program时这些Extra属性的恢复（目前保存时未裁剪Extra参数，可以直接从Program恢复到`runtime_attrs`中）同样也已经有单测覆盖到，在PR提交的早期对这些单测暴露的问题也已完成修复。

**Review by chenwhql — APPROVED** [review]

**Review by qingqing01 — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**